### PR TITLE
AEA-538 implement PyPI checks for satisfiability.

### LINKED ISSUE #1032: Check for PyPI dependenceis version consistency across packages
**Is your feature request related to a problem? Please describe.**
If two packages have conflicting PyPI dependencies, we don't detect that.

**Describe the solution you'd like**
Check that we always have compatible version specifiers across packages.

**Describe alternatives you've considered**
N/A

**Additional context**
N/A

### PR DESCRIPTION
**Pull request by marcofavorito**
## Proposed changes

Implement the PyPI satisfiability check for a set of version specifiers.

## Fixes

Fixes #1032 

## Types of changes

What types of changes does your code introduce to agents-aea?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

_Put an `x` in the boxes that apply._

- [ ] I have read the [CONTRIBUTING](../CONTRIBUTING.md) doc
- [ ] I am making a pull request against the `develop` branch (left side). Also you should start your branch off our `develop`.
- [ ] Lint and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have checked that code coverage does not decrease.
- [ ] I have checked that the documentation about the `aea cli` tool works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

## Further comments
n/a

### CODE REVIEW

File: aea/helpers/pypi_checker.py [hunk 1]
```diff
@@ -0,0 +1,297 @@
...
+
+        if result is False:
+            raise ValueError(
+                "Specifier set {} for package {} not satisfiable.".format(
+                    specifier_set, package_name
+                )
+            )
+        self._pypi_dependencies[package_name] = specifier_set
+
+    def _add_or_merge_with_existing_specifier_set(
```
Comment (cyenyxe) [inline aea/helpers/pypi_checker.py:263]: What is the use case for this method compared to `_merge_with_existing_specifier_set_if_any`?

File: aea/helpers/pypi_checker.py [hunk 2]
```diff
@@ -0,0 +1,297 @@
...
+        and merge with the new one.
+        Otherwise, return the new one.
+
+        :param package_name: the package name
+        :param new_specifier_set: the new specifier set.
+        :return: the merged specifier set, or only the new one.
+        """
+        if package_name in self._pypi_dependencies:
+            old_set = self._pypi_dependencies[package_name]
+            return and_(old_set, new_specifier_set)
```
Comment (cyenyxe) [inline aea/helpers/pypi_checker.py:295]: For consistency with the call from `_add_or_merge_with_existing_specifier_set`:
```suggestion
            return operator.and_(old_set, new_specifier_set)
```

File: aea/helpers/pypi_checker.py [hunk 3]
```diff
@@ -0,0 +1,297 @@
...
+    >>> _is_satisfiable(s1)
+    True
+
+    The specifier set "==1.0, >1.1" is not satisfiable:
+
+    >>> s1 = SpecifierSet("==1.0,>1.1")
+    >>> _is_satisfiable(s1)
+    False
+
+    Other examples:
```
[the remark this answers is not in the document]
Comment (marcofavorito) [reply aea/helpers/pypi_checker.py]: good point, will fix

**Comment by marcofavorito** [comment]
For now, I removed the `PyPIChecker` class to make things simpler. Otherwise, I would have to handle the removal of the PyPI dependencies of a certain AEA package, not only the addition of PyPI dependencies.

**Pull request merged**